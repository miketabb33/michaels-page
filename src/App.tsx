import React from 'react'
import styled from 'styled-components'
import Button from './components/mblocks/Button'
import Spacer from './components/mblocks/Spacer'
import { useTheme } from './context/ThemeContext'
import GlobalStyle from './styles/GlobalStyles'

const Screen = styled.div`
  height: 100vh;
`
const CardWrapper = styled.div`
  width: 300px;
  position: relative;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
`

const Portrait = styled.img`
  border-radius: 50%;
`

const App = () => {
  const { theme, toggleTheme } = useTheme()
  return (
    <>
      <GlobalStyle themes={theme} />
      <Screen>
        <CardWrapper>
          <Portrait src="./images/portrait.jpeg" />
          <Spacer size="medium" />
          <h3>Hi Shanna</h3>
          <p>Im Michael Tabb</p>
          <p>I love you! Have a great day!</p>
          <Spacer size="medium" />
          <Button onClick={toggleTheme}>New Look?</Button>
        </CardWrapper>
      </Screen>
    </>
  )
}

export default App
