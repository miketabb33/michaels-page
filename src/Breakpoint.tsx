export const desktopAndUp = '(min-width: 1000px)'
export const tabletAndUp = '(min-width: 640px)'
