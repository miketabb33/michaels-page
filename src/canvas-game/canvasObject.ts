import { Direction } from '../types/Direction'
import { Rect } from '../types/Rect'
import { Velocity } from '../types/Velocity'

export type CanvasObjectType = {
  rect: Rect
  color: string
  speed: number
  direction: Direction
}

export type CanvasObjectController = {
  getCanvasObject: () => CanvasObjectType
  move: (direction: Direction) => void
  changeDirection: (direction: Direction) => void
  addVelocity: (velocity: Velocity) => void
}

const canvasObject = (init: CanvasObjectType): CanvasObjectController => {
  const width = init.rect.size.width
  const height = init.rect.size.height
  let x = init.rect.position.x
  let y = init.rect.position.y
  const speed = init.speed
  const color = init.color
  let direction = init.direction

  const move = (direction: Direction) => {
    const velocity = velocityMap.get(direction)
    if (velocity) {
      addVelocity(velocity)
    }
  }

  const addVelocity = (velocity: Velocity) => {
    const xTranslation = velocity.x * velocity.speed
    const yTranslation = velocity.y * velocity.speed
    x += xTranslation
    y += yTranslation
  }

  const velocityMap = new Map<Direction, Velocity>([
    ['up', { x: 0, y: -1, speed }],
    ['right', { x: 1, y: 0, speed }],
    ['down', { x: 0, y: 1, speed }],
    ['left', { x: -1, y: 0, speed }],
    ['up right', { x: 0.5, y: -0.5, speed }],
    ['down right', { x: 0.5, y: 0.5, speed }],
    ['down left', { x: -0.5, y: 0.5, speed }],
    ['up left', { x: -0.5, y: -0.5, speed }],
  ])

  const changeDirection = (newDirection: Direction) => {
    direction = newDirection
  }

  const getCanvasObject = (): CanvasObjectType => {
    return {
      rect: {
        position: { x, y },
        size: { width, height },
      },
      color,
      speed,
      direction,
    }
  }

  return {
    getCanvasObject,
    move,
    changeDirection,
    addVelocity,
  }
}

export default canvasObject
