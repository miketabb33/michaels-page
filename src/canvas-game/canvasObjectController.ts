import { Rect } from './rect'
import { DirectionValue } from './types/DirectionValue'
import { Velocity } from './types/Velocity'

export type CanvasObjectShape = 'rectangle' | 'circle'

export type CanvasObject = {
  rect: Rect
  color: string
  velocity: Velocity
  shape: CanvasObjectShape
}

export type CanvasObjectController = {
  canvasObj: () => CanvasObject
  move: () => void
  changeDirection: (direction: DirectionValue) => void
  changeSpeed: (newSpeed: number) => void
}

const canvasObject = (init: CanvasObject): CanvasObjectController => {
  const width = init.rect.size.width
  const height = init.rect.size.height
  let x = init.rect.position.x
  let y = init.rect.position.y
  let speed = init.velocity.speed
  const color = init.color
  let directionalX = init.velocity.directionValue.x
  let directionalY = init.velocity.directionValue.y
  const shape = init.shape

  const move = () => {
    addVelocity({ speed, directionValue: { x: directionalX, y: directionalY } })
  }

  const addVelocity = (velocity: Velocity) => {
    const xTranslation = velocity.directionValue.x * velocity.speed
    const yTranslation = velocity.directionValue.y * velocity.speed
    x += xTranslation
    y += yTranslation
  }

  const changeDirection = (newDirection: DirectionValue) => {
    directionalX = newDirection.x
    directionalY = newDirection.y
  }

  const changeSpeed = (newSpeed: number) => {
    speed = newSpeed
  }

  const canvasObj = (): CanvasObject => {
    return {
      rect: {
        position: { x, y },
        size: { width, height },
      },
      color,
      velocity: {
        speed,
        directionValue: {
          x: directionalX,
          y: directionalY,
        },
      },
      shape,
    }
  }

  return {
    canvasObj,
    move,
    changeDirection,
    changeSpeed,
  }
}

export default canvasObject
