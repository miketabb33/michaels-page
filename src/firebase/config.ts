import { initializeApp } from 'firebase/app'
import { getFirestore } from 'firebase/firestore'

const firebaseConfig = {
  apiKey: 'AIzaSyDprPwOCjEwEYcbW4nuPxFWtEOJFMc6bek',
  authDomain: 'michael-website-19ed0.firebaseapp.com',
  projectId: 'michael-website-19ed0',
  storageBucket: 'michael-website-19ed0.appspot.com',
  messagingSenderId: '855765427298',
  appId: '1:855765427298:web:763f3885091a33b5231d3b',
}

const app = initializeApp(firebaseConfig)
export const db = getFirestore(app)
