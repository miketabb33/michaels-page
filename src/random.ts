export const random = (amount: number) => {
  return Math.floor(Math.random() * amount)
}
