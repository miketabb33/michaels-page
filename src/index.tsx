import React from 'react'
import { createRoot } from 'react-dom/client'
import App from './App'
import { StylesContextProvider } from './context/StylesContext'

const rootElement = document.getElementById('root')

if (!rootElement) throw new Error('Failed to find the root element')

const root = createRoot(rootElement)
root.render(
  <React.StrictMode>
    <StylesContextProvider>
      <App />
    </StylesContextProvider>
  </React.StrictMode>
)
