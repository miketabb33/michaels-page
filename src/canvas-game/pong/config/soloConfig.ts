import { random } from '../../../random'
import { Size } from '../../types/Size'
import { CanvasObject, CanvasObjectController } from '../../canvasObjectController'
import { ScoreResult } from '../../pongScore'
import { PongConfig } from './pongConfigs'
import { DirectionValue } from '../../types/DirectionValue'

const tuftsBlue = '#3C97D7'
const carrotOrange = '#D98F25'

export const getPongSoloConfig = (): PongConfig => {
  const canvasDimensionUnits = 1000
  return {
    pongBall: makePongBall(canvasDimensionUnits),
    playerPaddle: makePlayerPaddle(canvasDimensionUnits),
    opponentPaddle: makeOpponentPaddle(canvasDimensionUnits),
    canvasDimensionUnits,
    didFireFrame,
  }
}

const paddleSize: Size = { width: 150, height: 30 }

const makePlayerPaddle = (canvasUnits: number): CanvasObject => {
  return {
    rect: {
      size: paddleSize,
      position: {
        x: canvasUnits / 2 - paddleSize.width / 2,
        y: canvasUnits - paddleSize.height,
      },
    },
    color: tuftsBlue,
    velocity: {
      speed: 10,
      directionValue: {
        x: 0,
        y: 0,
      },
    },
    shape: 'rectangle',
  }
}

const makeOpponentPaddle = (canvasUnits: number): CanvasObject => {
  return {
    rect: {
      size: { ...paddleSize, width: canvasUnits },
      position: {
        x: 0,
        y: 0,
      },
    },
    color: tuftsBlue,
    velocity: {
      speed: 10,
      directionValue: {
        x: 0,
        y: 0,
      },
    },
    shape: 'rectangle',
  }
}

const makePongBall = (canvasUnits: number): CanvasObject => {
  const ballSize: Size = { width: 20, height: 20 }

  return {
    rect: {
      size: ballSize,
      position: {
        x: canvasUnits / 2 - ballSize.width / 2,
        y: canvasUnits / 2 - ballSize.height / 2 - 200,
      },
    },
    color: carrotOrange,
    velocity: {
      speed: levels[0].speed,
      directionValue: randomPaddleDirection(),
    },
    shape: 'circle',
  }
}

const randomPaddleDirection = (): DirectionValue => {
  const randomNum = random(6)

  if (randomNum === 0) return { x: 0.05, y: -0.95 }
  if (randomNum === 1) return { x: -0.5, y: -0.5 }
  if (randomNum === 2) return { x: 0.5, y: -0.5 }
  if (randomNum === 3) return { x: 0.05, y: 0.95 }
  if (randomNum === 4) return { x: -0.5, y: 0.5 }
  if (randomNum === 5) return { x: 0.5, y: 0.5 }
  return { x: 0, y: 0 }
}

const didFireFrame = (
  playerPaddle: CanvasObjectController,
  pongBall: CanvasObjectController,
  opponentPaddle: CanvasObjectController,
  scoreResult: ScoreResult
) => {
  const calculatedLevel = Math.floor(scoreResult.score / 4)
  const currentLevel = levels.find((pred) => pred.level === calculatedLevel) || levels[levels.length - 1]
  const currentBallSpeed = pongBall.canvasObj().velocity.speed

  if (currentBallSpeed !== currentLevel.speed) {
    pongBall.changeSpeed(currentLevel.speed)
    playerPaddle.changeSpeed(playerPaddle.canvasObj().velocity.speed - 0.5)
  }
}

type pongLevel = {
  speed: number
  level: number
}

const levels: pongLevel[] = [
  {
    level: 0,
    speed: 13,
  },
  {
    level: 1,
    speed: 14,
  },
  {
    level: 2,
    speed: 15,
  },
  {
    level: 3,
    speed: 16,
  },
  {
    level: 4,
    speed: 17,
  },
  {
    level: 5,
    speed: 18,
  },
  {
    level: 6,
    speed: 19,
  },
  {
    level: 7,
    speed: 20,
  },
]
