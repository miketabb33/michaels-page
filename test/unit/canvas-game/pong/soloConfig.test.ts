import { CanvasObjectController } from '../../../../src/canvas-game/canvasObjectController'
import { getPongSoloConfig } from '../../../../src/canvas-game/pong/config/soloConfig'
import { random } from '../../../../src/random'
import { makeCanvasObjectMock } from '../../__MOCKS__/canvasObject.mock'

jest.mock('../../../../src/random')

const RANDOM = random as jest.Mock

RANDOM.mockImplementation(() => 1)

const CHANGE_SPEED_PLAYER = jest.fn()
const CHANGE_SPEED_BALL = jest.fn()

const PLAYER_PADDLE = {
  canvasObj: () => ({ ...makeCanvasObjectMock(), velocity: { speed: 10 } }),
  changeSpeed: CHANGE_SPEED_PLAYER,
} as unknown as CanvasObjectController

beforeEach(jest.clearAllMocks)

describe('Solo Config', () => {
  it('should match snapshot', () => {
    expect(getPongSoloConfig()).toMatchSnapshot()
  })

  it('should not change speed when current ball speed and current level speed are the same', () => {
    const SCORE = 3
    const PONG_BALL = {
      canvasObj: () => ({ ...makeCanvasObjectMock(), velocity: { speed: 13 } }),
      changeSpeed: CHANGE_SPEED_BALL,
    } as unknown as CanvasObjectController

    const { didFireFrame } = getPongSoloConfig()
    didFireFrame({
      playerPaddle: PLAYER_PADDLE,
      pongBall: PONG_BALL,
      opponentPaddle: {} as CanvasObjectController,
      score: SCORE,
    })

    expect(CHANGE_SPEED_BALL).not.toHaveBeenCalled()
    expect(CHANGE_SPEED_PLAYER).not.toHaveBeenCalled()
  })
  it('should change speed when current ball speed and current level speed are NOT the same', () => {
    const SCORE = 5
    const PONG_BALL = {
      canvasObj: () => ({ ...makeCanvasObjectMock(), velocity: { speed: 13 } }),
      changeSpeed: CHANGE_SPEED_BALL,
    } as unknown as CanvasObjectController

    const { didFireFrame } = getPongSoloConfig()
    didFireFrame({
      playerPaddle: PLAYER_PADDLE,
      pongBall: PONG_BALL,
      opponentPaddle: {} as CanvasObjectController,
      score: SCORE,
    })

    expect(CHANGE_SPEED_BALL).toHaveBeenCalledWith(15)
    expect(CHANGE_SPEED_PLAYER).toHaveBeenCalledWith(9.5)
  })
})
