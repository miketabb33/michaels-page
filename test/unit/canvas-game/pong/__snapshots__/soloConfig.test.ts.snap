// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Solo Config should match snapshot 1`] = `
{
  "canvasDimensionUnits": 1000,
  "didFireFrame": [Function],
  "opponentPaddle": {
    "color": "#378ac5",
    "rect": {
      "position": {
        "x": 0,
        "y": 0,
      },
      "size": {
        "height": 30,
        "width": 1000,
      },
    },
    "shape": "rectangle",
    "velocity": {
      "directionValue": {
        "x": 0,
        "y": 0,
      },
      "speed": 10,
    },
  },
  "playerPaddle": {
    "color": "#378ac5",
    "rect": {
      "position": {
        "x": 425,
        "y": 970,
      },
      "size": {
        "height": 30,
        "width": 150,
      },
    },
    "shape": "rectangle",
    "velocity": {
      "directionValue": {
        "x": 0,
        "y": 0,
      },
      "speed": 10,
    },
  },
  "pongBall": {
    "color": "#b4771f",
    "rect": {
      "position": {
        "x": 490,
        "y": 290,
      },
      "size": {
        "height": 20,
        "width": 20,
      },
    },
    "shape": "circle",
    "velocity": {
      "directionValue": {
        "x": -0.5,
        "y": -0.5,
      },
      "speed": 13,
    },
  },
}
`;
